engine: default location for reading/writing jwt secrets (#297)
ethereum/execution-apis: src/engine/authentication.md

The `EL` and `CL` clients **MUST** accept a cli/config parameter: `jwt-secret`, which designates a file containing the hex-encoded 256 bit secret key to be used for verifying/generating JWT tokens.

If such a parameter _is_ given, but the file cannot be read, or does not contain a hex-encoded key of `256` bits, the client should treat this as an error: either abort the startup, or show error and continue without exposing the authenticated port.

> If such a parameter is not given, the client **MUST** attempt to read the secret from the default paths defined below, in the order they are listed.  If a secret is found, but the file cannot be read, does not contain a hex-encoded key of `256` bit, or is rejected by the other client, the client should continue searching. The CL **MAY** search other locations, such as default EL data directories.

Reviewer: > the client should continue searching

and ultimately abort if no valid file is found?
marginal worry it implies "search forever"

Author: Updated to more clearly state when an abort should occur.

> If such a parameter is not given, the client **MUST** attempt to read the secret from the default paths defined below, in the order they are listed.  If a secret is found, but the file cannot be read, does not contain a hex-encoded key of `256` bit, or is rejected by the other client, the client should continue searching. The CL **MAY** search other locations, such as default EL data directories.

Reviewer 2: As mentioned below, I'm not a fan of the **MUST** here.  A high security client may reasonably refuse to read secrets from disk and demand that they be read from a secure secret location in order to properly start.  Such a client shouldn't be in violation of the specification in that section.  I think a **SHOULD** would be appropriate here and address the need that I believe you are trying to address.

Author: I think a "high security" client should just flag this configuration with lots of warnings. If I knew all current clients would implement this if it is listed as **SHOULD**, I would be inclined to change it. But generally, I think this is an important UX improvement and we should force clients to follow suite.

Reviewer 2: If it is sufficiently useful to users, then all clients will implement it without being forced to.  I generally am pretty 👎 on sacrificing security for usability.  We should always try to find ways to improve usability, but never at the cost of security.

A "high security client" should not even start if it is configured in a way that is insecure.  This sort of client is valuable for businesses and enterprise shops where they want to make sure human error cannot result in a security compromise.  Those clients wouldn't be using this feature at all, as they would have their own key management system that injects the keys just-in-time (likely as part of whatever process is launching the client).

> If such a parameter is not given, the client **MUST** attempt to read the secret from the default paths defined below, in the order they are listed.  If a secret is found, but the file cannot be read, does not contain a hex-encoded key of `256` bit, or is rejected by the other client, the client should continue searching. The CL **MAY** search other locations, such as default EL data directories.

Reviewer 2: Recommend avoiding acronyms in specs when possible (I know they are used elsewhere, someone should fix those too).  Also, EL => EL Client *at least*.
```suggestion
If such a parameter is not given, the client **MUST** attempt to read the secret from the default paths defined below, in the order they are listed.  If a secret is found, but the file cannot be read, does not contain a hex-encoded key of `256` bit, or is rejected by the other client, the client should continue searching. The consensus layer client **MAY** search other locations, such as default execution layer client data directories.
```

Author: I will keep it as is and we can address these all at once in a separate PR (unless #298 is merged first).

> If no existing JWT is located, the client **MUST** generate such a token, valid for the duration of the execution, and store the hex-encoded secret on the filesystem.  The exact path is dependant on the host OS, see the default paths below.  If the client is unable to write the secret to an OS-specific path, it **MUST** fallback to writing to its own data directory.

Reviewer 2: I'm not a fan of requiring tokens be stored on disk.  A high security client build may do key management through an actually secure key management service and writing the key to disk would negate those security benefits.  Similarly, a high security client may entirely remove the option of automatically generating the token in order to protect against mis-configuration leading to a security breach and such a client shouldn't be considered out of spec.

Note: Reading over this document I see some other areas where the wording is too strong that I'll try to address in a different PR.

Reviewer 3: I think the intent of the JWT secret might play a role in the comment you made. The JWT secret is not meant to encrypt, it it meant to authenticate that the user meant to communicate with the given port. The vector that is being in case a user mistakenly exposes their engine API port to the outside world, that their EL will not be led astray. The authentication is purely meant to reduce the user error space, not to provide a secure medium of communication. 

In a scenario in which anyone has access to the disk and can read the JWT secret, we can assume the machine is completely compromised and protecting the user from an exposed port is not as high of a worry. Once you have disk level access, it would be relatively trivial to mess with the users EL, e.g: modifying their chain database or importing a new one.

Reviewer 2: > In a scenario in which anyone has access to the disk and can read the JWT secret, we can assume the machine is completely compromised and protecting the user from an exposed port is not as high of a worry. Once you have disk level access, it would be relatively trivial to mess with the users EL, e.g: modifying their chain database or importing a new one.

Situations where a host, network, or deployment may be partially compromised and writing the JWT to disk provides an escalation vector to the attacker:
* You may have read access to the device but not write access.
* You may have disk access to the location the JWTs are stored, but not to the location the application or DB is stored (especially likely if JWTs are written to a different directory structure as proposed here).
* You have access to a backup (historical read access) but do not have access to the live machine (no write access, no realtime read access).

In general, following a security in depth strategy should not cause someone to go out of spec compliance, and not writing secrets to disk is one layer of many that helps mitigate the chances of critical failure.

Author: I think this is mostly out-of-scope for this PR. The spec already requires the JWT token be written to a location both clients have access to. Almost certainly, this will be a relatively permissive location for other users to read. It is of course possible to manually configure that the keys are written to a more secure location. I think general security considerations with the JWT approach should be taken up in a separate PR.

Reviewer 2: > The spec already requires the JWT token be written to a location both clients have access to. 

Hmm, you are right.  I'll submit a PR to fix that as well.

### Default JWT secret locations

For Linux:

> * `$XDG_DATA_DIR/ethereum/engine/jwt.hex`

Reviewer 2: ```suggestion
* `$XDG_CONFIG_HOME/ethereum/engine/jwt.hex`
```
I believe this qualifies as configuration, not data, and thus should go into the config directory.  From XDK spec:
>  There is a single base directory relative to which user-specific data files should be written. This directory is defined by the environment variable $XDG_DATA_HOME.
> 
> There is a single base directory relative to which user-specific configuration files should be written. This directory is defined by the environment variable $XDG_CONFIG_HOME.

Reviewer 2: I'm not opposed to `XDG_CACHE_DIR`.  Given that this just needs to be set at startup and exist until the user launches the other client, cache seems quite reasonable as it isn't a big deal if this gets pruned by some cache cleanup mechanism.  That directory is also explicitly excluded from almost all backup systems which is a nice bonus.

Author: Updated.

* `$HOME/.local/share/ethereum/engine/jwt.hex`

For Mac:
* `$HOME/Library/Application Support/Ethereum/Engine/jwt.hex`

For Windows:

> * `{FOLDERID_RoamingAppData}/Ethereum/Engine/jwt.hex`

Author: This makes sense.

If such a parameter _is_ given, but the file cannot be read, or does not contain a hex-encoded key of `256` bits, the client **SHOULD** treat this as an error: either abort the startup, or show error and continue without exposing the authenticated port.

If such a parameter _is not_ given, the client **MUST** attempt to read the secret from the default paths defined below, in the order they are listed. If a secret is found, but the file cannot be read, does not contain a hex-encoded key of `256` bit, or is rejected by the other client, the client **MUST** continue searching until all default paths are exhausted. Consensus layer clients **MAY** search other locations, such as execution layer client data directories.

If no existing JWT is located, the client **SHOULD** generate such a token, valid for the duration of the execution, and store the hex-encoded secret on the filesystem. The exact path is dependant on the host OS, see the default paths below. If the client is unable to write the secret to an OS-specific path, it **MUST** fallback to writing to its own data directory.

### Default JWT secret locations

For Linux:

> * `$XDG_CACHE_DIR/ethereum/engine/jwt.hex`

Reviewer 4: Should be either of these instead:

- `XDG_CACHE_HOME/...` and `$HOME/.cache/...`
- `XDG_DATA_HOME/...` and `$HOME/.local/share/...`

Reviewer 2: Pretty sure CACHE is correct here because it can be regenerated as needed, and should *NOT* be included in any backup processes.
```suggestion
* `$XDG_CACHE_HOME/ethereum/engine/jwt.hex`
```
